src/shogun/lib/SGMatrix.cpp

```diff
@@ -229,49 +257,48 @@ bool SGMatrix<complex128_t>::is_symmetric()
 				return false;
 		}
 	}
+
 	return true;
 }
 
 template <class T>
-T SGMatrix<T>::max_single()
+T SGMatrix<T>::max_single() const
 {
-	T max=matrix[0];
-	for (int64_t i=1; i<int64_t(num_rows)*num_cols; ++i)
-	{
-		if (matrix[i]>max)
-			max=matrix[i];
-	}
+	REQUIRE(matrix!=nullptr, "The underlying matrix is null!\n");
+	REQUIRE(num_rows>0, "Number of rows is non-positive!\n");
+	REQUIRE(num_cols>0, "Number of cols is non-positive!\n");
 
-	return max;
+	return *std::max_element(matrix, matrix+size());
 }
 
 template <>
-complex128_t SGMatrix<complex128_t>::max_single()
+complex128_t SGMatrix<complex128_t>::max_single() const
 {
 	SG_SERROR("SGMatrix::max_single():: Not supported for complex128_t\n");
 	return complex128_t(0.0);
 }
 
 template <class T>
-SGMatrix<T> SGMatrix<T>::clone()
+SGMatrix<T> SGMatrix<T>::clone() const
 {
-	return SGMatrix<T>(clone_matrix(matrix, num_rows, num_cols),
-			num_rows, num_cols);
+	REQUIRE(matrix!=nullptr, "The underlying matrix is null!\n");
+	REQUIRE(num_rows>0, "Number of rows is non-positive!\n");
+	REQUIRE(num_cols>0, "Number of cols is non-positive!\n");
+
+	return SGMatrix<T>(clone_matrix(matrix, num_rows, num_cols), num_rows, num_cols);
 }
 
 template <class T>
 T* SGMatrix<T>::clone_matrix(const T* matrix, int32_t nrows, int32_t ncols)
 {
-	T* result = SG_MALLOC(T, int64_t(nrows)*ncols);
-	for (int64_t i=0; i<int64_t(nrows)*ncols; i++)
-		result[i]=matrix[i];
-
+	auto size=int64_t(nrows)*ncols;
+	T* result=SG_MALLOC(T, size);
+	std::copy(matrix, matrix+size, result);
```

Comment:
@lambday for future reference: if you would be doing this in a feature branch of shogun repo i could do these changes and merge myself the feature branch... but like this unfortunately i cannot... the only option would be to cherry pick your repo's feature branch ....... so yeah please since you have write access on the main repo just do the feature branch within that.